### PR TITLE
Formulates and implements properties for PVBUMP STS

### LINKED ISSUE #473: Formulate and implement properties for PVBUMP STS in Byron-Shelley
With the `PVBUMP` STS having no dependency to other STSes in the Byron-Shelley era specifications, formulate and implement a few properties for the STS.

### PR DESCRIPTION
**Pull request by mdimjasevic**
For the Byron-Shelley era, this patch formulates and implements three properties for the `PVBUMP` STS.

Closes #473.

### CODE REVIEW

File: byron/ledger/executable-spec/cs-ledger.cabal [hunk 1]
```diff
@@ -109,3 +109,36 @@ test-suite ledger-delegation-test
                "-with-rtsopts=-K1m -M50m"
   if (!flag(development))
     ghc-options: -Werror
+
+test-suite ledger-sts-test
```
Comment (dnadales) [inline byron/ledger/executable-spec/cs-ledger.cabal]: There is already a `test-suite ledger-delegation-test` section. Would it make sense to move the contents of the that section to this one (since the tests there are also testing the STS).

File: byron/ledger/executable-spec/test/Generator.hs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+-- | Generates a protocol version
+protVerGen :: Gen ProtVer
+protVerGen = do
+  pvMin  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  pvDiff <- Gen.integral (Range.linear (0 :: Natural) 20)
+  let pvMax = pvMin + pvDiff
+  pvAlt  <- Gen.integral (Range.linear (0 :: Natural) 100)
+  return $ ProtVer pvMin pvMax pvAlt
+
+-- | Generates a Double in the inclusive range [0, 1]
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: Why not using [`double`](http://hackage.haskell.org/package/hedgehog-1.0/docs/Hedgehog-Gen.html#v:double)?

File: byron/ledger/executable-spec/test/Generator.hs [hunk 3]
```diff
@@ -0,0 +1,109 @@
+module Generator where
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: We should move the functions here to modules that have a more specific name.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import           Data.Word (Word64)
+import           Hedgehog
+import qualified Hedgehog.Gen    as Gen
+import qualified Hedgehog.Range  as Range
+import           Ledger.Core (Slot(..), SlotCount(..), BlockCount(..))
+import           Ledger.GlobalParams (k)
+import           Ledger.Update (ProtVer(..), PParams(..))
+import           Numeric.Natural (Natural)
+
+-- | Generates a slot within the given bound
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: Maybe move this to `Ledger.Core.Generator`?

File: byron/ledger/executable-spec/test/Generator.hs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+slotGen :: Word64 -> Word64 -> Gen Slot
+slotGen lower upper = do
+  Slot <$> Gen.word64 (Range.linear lower upper)
+
+-- | Generates a protocol version
+protVerGen :: Gen ProtVer
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: We need to revisit this: in @edsko's talk we saw that monadic binds introduced unwanted dependencies among the generated data, so we might want to use the `Applicative` interface here.

Also in the protocol version the second number is the "major" version and not the "max", so since there are no dependencies between the generated data we can simply use `<$>` and `<*>`

File: byron/ledger/executable-spec/test/Generator.hs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+doubleGen :: Gen Double
+doubleGen = do
+  b <- Gen.word64 (Range.linear 1 hi)
+  a <- Gen.word64 (Range.linear 0 b )
+  return $ (fromIntegral a) / (fromIntegral b)
+ where
+  hi = 2 `shiftL` 48 :: Word64
+
+-- | Generates valid protocol parameters
+pparamsGen :: Gen PParams
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: Here we need to use the `Applicative` interface whenever possible, I have a similar generator [here](https://github.com/input-output-hk/cardano-ledger-specs/pull/490/files#diff-3836797915c9b23257caef6f9996c602R181).

Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: And even the generator I linked won't shrink properly, but at least we should get rid of the monadic dependencies insofar possible.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 4 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: So we might want to have these generators in the `src` directory so that they can be used in the conformance testing at `cardano-ledger`.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      maxPropSz
+      bkSgnCntT
+      bkSlotsPerEpoch
+      upTtl
+      scriptVersion
+      cfmThd
+      upAdptThd
+      stableAfter
+
+data PVBUMPEnv = PVBUMPEnv
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: This should be type `PVBUMPEnv = Environment PVBUMP` (or just use `Environment PVBUMP`).

**Comment by dnadales** [comment]
Just as a general comment, I think we should avoid the use of monadic binds in our generators whenever possible.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    s   <- slotGen 0 100000
+    pv  <- protVerGen
+    pps <- pparamsGen
+    return (s, (pv, pps))
+
+-- | Generates an environment for the PVBUMP STS with an empty list of
+-- updates
+pvbumpEmptyListEnvGen :: Gen PVBUMPEnv
+pvbumpEmptyListEnvGen =
+  pvbumpEnvGen >>= (\(PVBUMPEnv (s_n, _)) -> pure $ PVBUMPEnv (s_n, []))
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: We do not need to generate the list in `pvbumpEnvGen` here. Maybe it is enough to use:

```haskell
(, []) <$> slotGen 0 blah 
```

File: byron/ledger/executable-spec/test/Generator.hs (hunk 7 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Actually I'm thinking of changing the type in the STS itself. The reason is that wrappers like these help in both communicating to the developer what this is about (compared to a pair or a product in general) and also helps to avoid bugs due to otherwise hard to distinguish types.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 8 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Thanks! I've changed it in this manner and will push it later.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 2 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Because its upper bound is exclusive, while I need it inclusive here.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 3 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: I asked in your PR #490, but I'll do it here as it is related: how about extracting those generators that you have from a class instance and making them available at the top level? I guess as we go we'll figure out how to organize them into modules, but your suggestion to follow `src` module names makes sense to me.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 7 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: >  I'm thinking of changing the type in the STS itself

I'd suggest not to do that in this PR. Also I don't know what do you plan on changing the type to. For the chain executable spec I thought about using tuples, since this results in the formal and executable spec having a very similar appearance. However compiler errors are harder to debug.

Whatever you do, please include @nc6 and me in the initial design.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 3 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: > how about extracting those generators that you have from a class instance and making them available at the top level? 

Sure, if they are generators that other modules could use we should expose them.

> I guess as we go we'll figure out how to organize them into modules,

I'd rather try to distribute generators to modules now, otherwise we might run the risk of people re-writing these generators because they didn't find them.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+      upAdptThd
+      stableAfter
+
+data PVBUMPEnv = PVBUMPEnv
+  ( Slot
+  , [(Slot, (ProtVer, PParams))]
+  ) deriving Show
+
+-- | Generates an environment for the PVBUMP STS
+pvbumpEnvGen :: Gen PVBUMPEnv
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: So to clarify what I said above, regardless of what changes we make in the future, here we could simply write:

```haskell
pvbumpEnvGen :: Gen (Environment PVBUMP)

```

Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: So no need for new `data` or `type` aliases.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+pvbumpEnvGen = do
+  s_n <- slotGen 0 100000
+  l   <- listGen
+  return $ PVBUMPEnv (s_n, l)
+ where
+  listGen :: Gen [(Slot, (ProtVer, PParams))]
+  listGen = do
+    Gen.list (Range.linear 0 3) inOneSlot
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: Why do we generate lists of at most 3 elements?

File: byron/ledger/executable-spec/test/Generator.hs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    pps <- pparamsGen
+    return (s, (pv, pps))
+
+-- | Generates an environment for the PVBUMP STS with an empty list of
+-- updates
+pvbumpEmptyListEnvGen :: Gen PVBUMPEnv
+pvbumpEmptyListEnvGen =
+  pvbumpEnvGen >>= (\(PVBUMPEnv (s_n, _)) -> pure $ PVBUMPEnv (s_n, []))
+
+pvbumpBeginningsEnvGen :: Gen PVBUMPEnv
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: I'm trying to understand why this generator is called like this. Maybe add a comment?

File: byron/ledger/executable-spec/test/Generator.hs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+pvbumpBeginningsEnvGen = do
+  s_n <- slotGen 0 $ 2 * (unBlockCount k)
+  env <- pvbumpEnvGen
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Generator.hs]: Here it seems we're also using `pvbumpEnvGen` only to generate a ` [(Slot, (ProtVer, PParams))]`. So maybe we can have a generator for this, and remove `pvbumpEnvGen`?

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+-- Given an empty list of mappings from a slot to a pair of a protocol
+-- version and protocol parameters (this is the fads part of the
+-- environment), the PVBUMP STS results in the same state it started
+-- in.
+emptyPVUpdate :: Property
+emptyPVUpdate = property $ do
+  PVBUMPEnv env  <- forAll pvbumpEmptyListEnvGen
+  PVBUMPState st <- forAll pvbumpStateGen
+  sig            <- forAll pvbumpSignalGen
+  case applySTS @PVBUMP $ TRC (env, st, sig) of
```
Comment (dnadales) [inline byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Why generating the signal instead of simply writing:

```haskell
  case applySTS @PVBUMP $ TRC (env, st, ()) of
```

File: byron/ledger/formal-spec/properties.tex [hunk 14]
```diff
@@ -156,3 +156,76 @@ \subsection{UTxO Properties}
   $$
 
 \end{property}
+
+\subsection{PVBUMP STS Properties}
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: It'd be nice to be consistent with the name of the other sections, we either make all `$LABEL STS Properties` or `$LABEL Properties`. I'd prefer the latter, since it is inside a section named "Transition System Properties".

Also we might want to add this into a sub-sub section, inside a section named "Update Properties" or something like that.

File: byron/ledger/formal-spec/properties.tex [hunk 15]
```diff
@@ -156,3 +156,76 @@ \subsection{UTxO Properties}
...
+\label{sec:pvbump-properties}
+
+Property~\ref{prop:pvbump-no-proposals} states that when no proposals are
+available, the system remains in the same state it started with.
+
+\begin{property}[PVBUMP on no proposals]\label{prop:pvbump-no-proposals}
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: In these properties we're not writing an inference rule, so I'd suggest to rephrase this as something like:

image

Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: Maybe we need to find a better label for the property. What about "no change on empty future adoptions" or something like that ...

File: byron/ledger/formal-spec/properties.tex [hunk 16]
```diff
@@ -156,3 +156,76 @@ \subsection{UTxO Properties}
...
+         \right)
+       }
+     }
+\end{property}
+
+Property~\ref{prop:pvbump-early-on} states that early on in the lifetime
+of the blockchain the PVBUMP system remains in the same state it started with.
+
+\begin{property}[PVBUMP early on]\label{prop:pvbump-early-on}
+  \textrm{For all current slots that have an index less than or equal to
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: Maybe this prose can be outside of the property.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 5 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Fixed.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 6 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Fixed.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 7 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Sure, I'll leave it that way in `tests` only.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 9 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: I use a slightly different approach here, as explained earlier.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 10 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: For no particular reason. What number do you think should go there instead?

Comment (dnadales) [reply byron/ledger/executable-spec/test/Generator.hs]: I don't know but it seems like a low number. That's why I asked.

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 13 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Obviously either would work. The way it is now shows that these values need not be hard-coded. If you're keen on seeing what you proposed, I'll change it.

File: byron/ledger/executable-spec/cs-ledger.cabal [hunk 17]
```diff
@@ -109,3 +109,36 @@ test-suite ledger-delegation-test
...
+  hs-source-dirs: test
+  main-is: STS.hs
+  exposed-modules: Generator
+                 , Ledger.Pvbump.Properties
+  other-modules:   Generator
```
Comment (ruhatch) [inline byron/ledger/executable-spec/cs-ledger.cabal]: These should only be in one section, exposed or other

File: byron/ledger/executable-spec/cs-ledger.cabal [hunk 18]
```diff
@@ -109,3 +109,36 @@ test-suite ledger-delegation-test
...
+               , small-steps
+  ghc-options: -Wall
+               -Wcompat
+               -Wincomplete-record-updates
+               -Wincomplete-uni-patterns
+               -Wredundant-constraints
+               -- We set a bound here so that we're alerted of potential space
+               -- leaks in our generators (or test) code.
+               --
+               -- The 50 megabytes heap bound was determined ad-hoc.
```
Comment (ruhatch) [inline byron/ledger/executable-spec/cs-ledger.cabal]: :+1:

File: byron/ledger/executable-spec/test/Generator.hs [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+protVerGen =
+  (\(a, b) alt -> ProtVer a b alt) <$>
+    minMaj <*>
+    Gen.integral (Range.linear (0 :: Natural) 100)
+ where
+  minMaj :: Gen (Natural, Natural)
+  minMaj = do
+    pvMin  <- Gen.integral (Range.linear (0 :: Natural) 100)
+    pvDiff <- Gen.integral (Range.linear (0 :: Natural) 20)
+    pure (pvMin, pvMin + pvDiff)
```
Comment (ruhatch) [inline byron/ledger/executable-spec/test/Generator.hs]: I don't quite understand what this is doing - what's the relationship being captured here?

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 13 above)
Comment (dnadales) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: `PVBUMP` signals can only be `()`. Personally I prefer to avoid writing unnecessary code, but do whatever you think is sensible.

File: byron/ledger/executable-spec/test/Generator.hs [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+
+-- | Generates a Double in the inclusive range [0, 1]
+doubleGen :: Gen Double
```
Comment (ruhatch) [inline byron/ledger/executable-spec/test/Generator.hs]: Is [`Gen.double`](http://hackage.haskell.org/package/hedgehog-1.0/docs/Hedgehog-Internal-Gen.html#v:double) not enough here?

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+-- 
+-- Given an empty list of mappings from a slot to a pair of a protocol
+-- version and protocol parameters (this is the fads part of the
+-- environment), the PVBUMP STS results in the same state it started
+-- in.
+emptyPVUpdate :: Property
+emptyPVUpdate = property $ do
+  PVBUMPEnv env  <- forAll pvbumpEmptyListEnvGen
+  PVBUMPState st <- forAll pvbumpStateGen
+  sig            <- forAll pvbumpSignalGen
```
Comment (ruhatch) [inline byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: To improve shrinking it would be better to use `Applicative` here, so
```
judgementContext <- forAll $ fmap TRC $ (,,) <$> pvbumpEmptyListEnvGen <*> pvbumpStateGen <*> pvbumpSignalGen
```

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+    Right st' -> st === st'
+    Left _    -> failure
+
+-- Property #2 for the PVBUMP STS
+--
+-- For s_n <= 2 * k, the resulting state is the same as the state the
+-- system started in.
+beginningsNoUpdate :: Property
+beginningsNoUpdate = property $ do
+  PVBUMPEnv env  <- forAll pvbumpBeginningsEnvGen
```
Comment (ruhatch) [inline byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Similar here to avoid shrinking dependencies

File: byron/ledger/executable-spec/cs-ledger.cabal (hunk 17 above)
Comment (nc6) [reply byron/ledger/executable-spec/cs-ledger.cabal]: Expose everything, particularly in testing libraries.

File: byron/ledger/executable-spec/cs-ledger.cabal (hunk 1 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/cs-ledger.cabal]: Sure, that's completely doable, though the test suite's name would be misleading in such a case.

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 21 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: I agree with the general idea, but when I do that then many `Show` instances are needed, in particular for `TRC` and all environments, states, and signals. Does it sound reasonable to add the instances?

File: byron/ledger/executable-spec/test/Generator.hs (hunk 20 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: No, because `Gen.double` does not have an inclusive upper bound.

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 13 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: `()` it is.

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 22 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: @ruhatch , please help here, the issues is the same as in https://github.com/input-output-hk/cardano-ledger-specs/pull/488/files#r286938596

File: byron/ledger/executable-spec/test/Generator.hs (hunk 3 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Done! Ledger update generators are now in `src` in the `Ledger.Update.Generator` module.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 4 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Done!

File: byron/ledger/executable-spec/cs-ledger.cabal (hunk 1 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/cs-ledger.cabal]: I merged the two test suites into one.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 9 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: I implemented what you suggested.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 10 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: If so, feel free to suggest a higher number and I'll be happy to include it.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 11 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: It has such a comment now.

File: byron/ledger/executable-spec/test/Generator.hs (hunk 12 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: I'll leave `pvbumpEnvGen` for now as I'm still to implement one property for `PVBUMP`, where this generator might be useful. I'll see then if it's needed.

File: byron/ledger/formal-spec/properties.tex (hunk 14 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: I addressed both of your comments. Thanks!

File: byron/ledger/executable-spec/test/Generator.hs (hunk 19 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: Oh, this was completely off. It is fixed now. Thanks for spotting it!

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 21 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Fixed!

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 22 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Fixed!

File: byron/ledger/formal-spec/properties.tex (hunk 15 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Thanks! Done!

Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Fixed!

File: byron/ledger/formal-spec/properties.tex (hunk 16 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Fixed!

File: byron/ledger/executable-spec/test/Generator.hs (hunk 10 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Generator.hs]: I parameterised this so it can be set to anything. There are examples that generate a list of 10 elements.

File: byron/semantics/executable-spec/src/Control/State/Transition.hs [hunk 23]
```diff
@@ -52,6 +52,12 @@ newtype IRC sts = IRC (Environment sts)
 -- | Context available to transition rules.
 newtype TRC sts = TRC (Environment sts, State sts, Signal sts)
 
+deriving instance
```
Comment (ruhatch) [inline byron/semantics/executable-spec/src/Control/State/Transition.hs]: I believe you could just write `deriving (Show)` on the `newtype` declaration

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+lastProposal = property $ do
+  judgementContext <-
+    forAll $ fmap TRC $ (,,)
+      <$> pvbumpAfter2kEnvGen
+      <*> pvbumpStateGen
+      <*> pure ()
+  let
+    TRC ((s_n, fads), _, _) = judgementContext
+    s = fromMaybe
+      undefined -- this case never happens because used generators
```
Comment (ruhatch) [inline byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: It would be good to throw a descriptive error here, otherwise `undefiend` could be confusing if someone breaks your expectations in the future

File: byron/ledger/formal-spec/properties.tex [hunk 25]
```diff
@@ -210,5 +210,100 @@ \subsection{UTxO Properties}
...
+\subsection{Update Properties}
+\label{sec:update-properties}
+
+\subsubsection{PVBUMP Properties}
+\label{sec:pvbump-properties}
+
+Property~\ref{PVBUMP-empty-future-adopt} states that when no proposals are
+available, the system remains in the same state it started with.
 
+\begin{property}[PVBUMP no change on empty future adoptions]\label{PVBUMP-empty-future-adopt}
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: Nice. Here we need to universally quantify on `sn`, `pv`, and `pps`.

Also if you put the predicate that follows the `such that` part you can avoid awkward line breaks:

image

File: byron/ledger/formal-spec/properties.tex [hunk 26]
```diff
@@ -210,5 +210,100 @@ \subsection{UTxO Properties}
...
+      \var{pv}, \var{pps}\\
+    \end{array}
+  \right)
+  $$
+\end{property}
+
+Property~\ref{prop:pvbump-early-on} states that early on in the lifetime
+of the blockchain the PVBUMP system remains in the same state it started with.
+
+\begin{property}[PVBUMP early on]\label{prop:pvbump-early-on}
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: Same here: we need to quantify over `sn`, `fads`, `pv`, and `pps`.

File: byron/ledger/formal-spec/properties.tex [hunk 27]
```diff
@@ -210,5 +210,100 @@ \subsection{UTxO Properties}
...
 \end{property}
+
+Property~\ref{prop:pvbump-last-proposal} states that for all current slots
+with an index $2 \cdot k$ or higher and a proposal list with at least one
+proposal from a slot greater than $2 \cdot k$, the resulting state is
+exclusively determined by the last pair from the list comprising a new
+protocol version and protocol parameters that is after a slot with an index
+$2 \cdot k$.
+
+\begin{property}[PVBUMP last proposal]\label{prop:pvbump-last-proposal}
```
Comment (dnadales) [inline byron/ledger/formal-spec/properties.tex]: We need to add quantifications here as well. And I think we don't need the word `current`, they're just slots.

File: byron/semantics/executable-spec/src/Control/State/Transition.hs (hunk 23 above)
Comment (mdimjasevic) [reply byron/semantics/executable-spec/src/Control/State/Transition.hs]: This would impose the Show constraint on `Environment sts`, `State sts` and `Signal sts`, which isn't there at the moment. This way it's flexible and not imposed globally.

File: byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs (hunk 24 above)
Comment (mdimjasevic) [reply byron/ledger/executable-spec/test/Ledger/Pvbump/Properties.hs]: Fixed! Will push it in a moment.

File: byron/ledger/formal-spec/properties.tex (hunk 25 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: I'm not sure I'm following. Can you explain again?

Comment (dnadales) [reply byron/ledger/formal-spec/properties.tex]: By quantifying I mean something like:

```
For all `sn`, `pv`, and `pps` such that

    ... blah ...

we have 

   ... the transition ...
```

Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Done!

File: byron/ledger/formal-spec/properties.tex (hunk 27 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Fixed!

File: byron/ledger/formal-spec/properties.tex (hunk 26 above)
Comment (mdimjasevic) [reply byron/ledger/formal-spec/properties.tex]: Fixed!
